enhancements/microshift/optionally-deploy-csi-plugin.md

```diff
@@ -0,0 +1,209 @@
+---
+title: optionally-deploy-csi-plugin
+authors:
+  - "@copejon"
+reviewers:
+  - "@pacevedom: MicroShift team-lead"
+  - "@jerpeter1, Edge Enablement Staff Engineer"
+  - "@jakobmoellerdev Edge Enablement LVMO SME"
+  - "@suleymanakbas91,  Edge Enablement LVMO SME"
+approvers:
+  - "@jerpeter1, Edge Enablement Staff Engineer"
+api-approvers:
+  - "None"
+creation-date: 2024-06-10
+last-updated: 2024-06-11
+tracking-link:
+  - https://issues.redhat.com/browse/OCPSTRAT-856
+see-also:
+  - "enhancements/microshift/microshift-default-csi-plugin.md"
+---
+
+# Optionally Deploy CSI Plugin
+
+## Summary
+
+MicroShift is a small form-factor, single-node OpenShift targeting IoT and Edge Computing use cases characterized by
+tight resource constraints, unpredictable network connectivity, and single-tenant workloads. See
+[kubernetes-for-devices-edge.md](./kubernetes-for-device-edge.md) for more detail.
+
+Out of the box, MicroShift includes the LVMS CSI driver and CSI snapshotter. The LVM Operator is not itself packaged
+with the platform, in adherence with the project's [guiding principles](./kubernetes-for-device-edge.md#goals). See the
+[MicroShift Default CSI Plugin](microshift-default-csi-plugin.md) proposal for a more in-depth explanation of the
+storage provider and reasons for its integration into MicroShift. Configuration of the driver is exposed via a config
+file at /etc/microshift/lvmd.yaml. The manifests required to deploy and configure the driver and snapshotter are baked
+into the MicroShift binary during compilation and are deployed during runtime. This means that even if a user wanted to
+disable the driver and/or snapshotter, they would be deployed again when the service restarts. Additionally, the LVMS an
+CSI containers required to run LVMS are always packaged with the product and thus will always consume a certain amount
+of storage space devices where MicroShift is deployed.
+
+## Motivation
+
+Not all users may persistent storage provider or volume snapshot capabilities for their MicroShift deployments and
+should be enabled to choose whether to deploy the driver and / or snapshotter. This will afford resource-conscious users
+the opportunity to better tune their MicroShift deployment to their hardware requirements.
+
+### User Stories
+
+- A user deploys MicroShift with the LVMS driver and CSI snapshot controller.
+- A user deploys MicroShift with the LVMS driver only.
+- A user deploys MicroShift with neither the LVMS driver nor the CSI snapshot controller.
+
+### Goals
+
+- Provide the LVMS CSI driver and LVMS CSI snapshot controller as optional components of MicroShift
+- Reuse the optional-installation pattern implemented by MicroShift for the [Multus CNI
+  Plugin](./multus-cni-for-microshift.md).
```

Comment:
LVMS / TopoLVM has custom logic paths in the microshift binary right now. If we want to have them in a seperate rpm, we need a new binary that is able to run as root and run these logic paths in its stead, especially for configuring lvmd.yaml with defaults. This will get only more pressing once we incorporate the hot reloading implementation in the LVMS migration from TopoLVM. Whats the idea to build that logic or run that somewhere?

Reply:
I believe this is something that would be well suited to a %post-install in the rpm spec. The logic is relatively simple and is mostly process calls wrapped with go code.  This script should also include logic to prevent stomping on existing config file during install so that user configs aren't lost after an upgrade/downgrade.

Reply:
With the migration to LVMS, to allow the hot reloading mechanism of lvmd.yaml we need to actively watch and copy over /etc/microshift/lvmd.yaml to /var/lib/microshift/lvms/lvmd.yaml

Also we currently have on every restart a copying of that file, that behavior would need to change if we move to post-install. 

Not generally opposed, just want to make sure we

- dont loose the ability to enable hotreloading in lvms for microshift (allows customers to reconfigure storage without restarting)
- dont break compatibility for ostree-style commits when upgrading

Reply:
I think this is compatible with the hotreloading behavior. The post-install script would be applying the changes to the /etc/microshift/lvmd.yaml (if not already present). So if a user were to install LVMS on a running cluster, without providing an lvmd.yaml file, the hot-reload should detect the new file and perform as expected.